### PR TITLE
[mcd] Initial contribution

### LINKED ISSUE #12201: Add new MCD Binding
The new binding is meant to send data to the MCD (managing care digital) cloud application by C&S Computer und Software GmbH.

### PR DESCRIPTION
**Pull request by simon-dengler**
Signed-off-by: Simon Dengler 

# [MCD] Initial Contribution

# Description

This contribution adds the MCD Binding to the project.
No changes to previously existing bindings have been made.

Fixes #12201

### CODE REVIEW

**Comment by simon-dengler** [comment]
@fwolter here is finally a clean new pull request on the new MCD binding (issue #12202 )
I have worked through all your feedback from the last submit. Thanks for your helpful hints!
best regards
Simon

**Review by fwolter — CHANGES_REQUESTED** [review]
There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html` after building it with `mvn clean install`.

You could create the default translations file by running `org.openhab.core.tools:i18n-maven-plugin:3.3.0:generate-default-translations` in your binding's directory. Then, [Crowdin](https://crowdin.com/project/openhab-addons) will pick up the English "translation" and users can help to translate your binding in Crowdin. If you want you can use Crowdin, too to translate your binding as soon as this PR is merged.

File: bundles/org.openhab.binding.mcd/README.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+
+## Discovery
+
+Discovery is not supported.
+
+
+## Binding Configuration
+
+No binding configuration required.
```
[the remark this answers is not in the document]
Comment (simon-dengler) [reply bundles/org.openhab.binding.mcd/README.md:22]: Done

File: bundles/org.openhab.binding.mcd/README.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+| email     | Email of account                   |
+| password  | valid password for the given email |
+
+### MCD Sensor Thing
+
+Each sensor thing (`mcdSensor`) needs to be configured with the identical serial number, that is assigned to this sensor in MCD. 
+
+| parameter      | description                        |
+|----------------|------------------------------------|
+| parent bridge  | parent MCD bridge is required      |
```
[the remark this answers is not in the document]
Comment (simon-dengler) [reply bundles/org.openhab.binding.mcd/README.md:44]: Done

File: bundles/org.openhab.binding.mcd/README.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+| serial number  | serial number of the sensor in MCD |
```
[the remark this answers is not in the document]
Comment (simon-dengler) [reply bundles/org.openhab.binding.mcd/README.md:45]: Fixed this and checked the bridge's parameters, too

File: bundles/org.openhab.binding.mcd/src/main/feature/feature.xml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<features name="org.openhab.binding.mcd-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
+	<repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${ohc.version}/xml/features</repository>
+
+	<feature name="openhab-binding-mcd" description="MCD Binding" version="${project.version}">
```
[the remark this answers is not in the document]
Comment (simon-dengler) [reply bundles/org.openhab.binding.mcd/src/main/feature/feature.xml:5]: Done

File: bundles/org.openhab.binding.mcd/src/main/java/org/openhab/binding/mcd/internal/handler/McdBridgeHandler.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        if (localConfig != null) {
+            try {
+                Request request = httpClient.newRequest("https://cunds-syncapi.azurewebsites.net/token")
+                        .method(HttpMethod.POST).header(HttpHeader.CONTENT_TYPE, "application/x-www-form-urlencoded")
+                        .header(HttpHeader.HOST, "cunds-syncapi.azurewebsites.net")
+                        .header(HttpHeader.ACCEPT, "application/json");
+                /*
+                 * String mail = config.getUserEmail();
+                 * String pwd = config.getUserPassword();
+                 */
```
[the remark this answers is not in the document]
Comment (simon-dengler) [reply bundles/org.openhab.binding.mcd/src/main/java/org/openhab/binding/mcd/internal/handler/McdBridgeHandler.java:111]: Done

File: bundles/org.openhab.binding.mcd/src/main/java/org/openhab/binding/mcd/internal/handler/McdBridgeHandler.java [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+                            case 200:
+                                if (content != null && content.has("access_token")) {
+                                    updateStatus(ThingStatus.ONLINE);
+                                    accessToken = content.get("access_token").getAsString();
+                                    expiresIn = content.get("expires_in").getAsInt();
+                                    long delay = ((long) expiresIn) * 1000L - 60000L;
+                                    Runnable task = () -> {
+                                        logMeIn();
+                                    };
+                                    scheduler.schedule(task, delay, TimeUnit.MILLISECONDS);
```
[the remark this answers is not in the document]
Comment (simon-dengler) [reply bundles/org.openhab.binding.mcd/src/main/java/org/openhab/binding/mcd/internal/handler/McdBridgeHandler.java:132]: Done

File: bundles/org.openhab.binding.mcd/src/main/java/org/openhab/binding/mcd/internal/handler/SensorThingHandler.java [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+            // check for the right channel id
+            if (channelId.equals(SEND_EVENT)) {
+                String commandString = command.toString();
+                int sensorEventId = SensorEventDef.getSensorEventId(commandString);
+                if (sensorEventId < 1 || sensorEventId > maxSensorEventId) {
+                    // check, if an id is passed as number
+                    try {
+                        sensorEventId = Integer.parseInt(commandString);
+                        if (sensorEventId < 1 || sensorEventId > maxSensorEventId) {
+                            logger.info("Invalid Command!");
```
[the remark this answers is not in the document]
Comment (simon-dengler) [reply bundles/org.openhab.binding.mcd/src/main/java/org/openhab/binding/mcd/internal/handler/SensorThingHandler.java:99]: Changed all

File: bundles/org.openhab.binding.mcd/src/main/java/org/openhab/binding/mcd/internal/handler/SensorThingHandler.java [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Cannot access config data.");
+        }
+        McdBridgeHandler localMcdBridgeHandler = mcdBridgeHandler;
+        if (localMcdBridgeHandler != null) {
+            // if (localMcdBridgeHandler.getAccessToken() != null) {
+            updateStatus(ThingStatus.ONLINE);
+            // } else {
+            // updateStatus(ThingStatus.OFFLINE);
+            // }
```
[the remark this answers is not in the document]
Comment (simon-dengler) [reply bundles/org.openhab.binding.mcd/src/main/java/org/openhab/binding/mcd/internal/handler/SensorThingHandler.java:133]: Removed all commented code

File: bundles/org.openhab.binding.mcd/src/main/java/org/openhab/binding/mcd/internal/handler/SensorThingHandler.java [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+                            /*
+                             * eventDef = new JsonObject();
+                             * for (JsonElement elem : eventDefArray) {
+                             * JsonObject obj = elem.getAsJsonObject();
+                             * // fill object with pairs of key (name of sensor event) and value (id of
+                             * event)
+                             * eventDef.addProperty(obj.get("NameApiSensorEventDef").getAsString(),
+                             * obj.get("IdApiSensorEventDef").getAsInt());
+                             * }
+                             */
```
[the remark this answers is not in the document]
Comment (simon-dengler) [reply bundles/org.openhab.binding.mcd/src/main/java/org/openhab/binding/mcd/internal/handler/SensorThingHandler.java:169]: Done

File: bundles/org.openhab.binding.mcd/src/main/java/org/openhab/binding/mcd/internal/handler/SensorThingHandler.java [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+    private void sendSensorEvent(@Nullable String serialNumber, int sensorEventDef) {
+        try {
+            McdBridgeHandler localMcdBridgeHandler = mcdBridgeHandler;
+            if (localMcdBridgeHandler != null) {
+                String accessToken = localMcdBridgeHandler.getAccessToken();
+                Date date = new Date();
+                String dateString = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ").format(date);
+                Request request = httpClient.newRequest("https://cunds-syncapi.azurewebsites.net/api/ApiSensor")
+                        .method(HttpMethod.POST).header(HttpHeader.CONTENT_TYPE, "application/json")
+                        // .header(HttpHeader.HOST, "cunds-syncapi.azurewebsites.net")
```
[the remark this answers is not in the document]
Comment (simon-dengler) [reply bundles/org.openhab.binding.mcd/src/main/java/org/openhab/binding/mcd/internal/handler/SensorThingHandler.java:415]: Removed it

File: bundles/org.openhab.binding.mcd/src/main/resources/OH-INF/thing/thing-types.xml [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+				<description>Password for your MCD account.</description>
+			</parameter>
+		</config-description>
+	</bridge-type>
+
+	<thing-type id="mcdSensor">
+		<supported-bridge-type-refs>
+			<bridge-type-ref id="mcdBridge"/>
+		</supported-bridge-type-refs>
+		<label>MCD Sensor Thing</label>
```
[the remark this answers is not in the document]
Comment (simon-dengler) [reply bundles/org.openhab.binding.mcd/src/main/resources/OH-INF/thing/thing-types.xml:28]: Done

File: bundles/org.openhab.binding.mcd/README.md [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+| channel     | type   | description                                   |
+|-------------|--------|-----------------------------------------------|
+| lastEvent | String | shows the last event that was sent with date and time |
+| sendEvent | String | stateless channel for sending events to the API |
+
+The channel `sendEvents` accepts valid Sensor Event Definitions as well as the corresponding ID. The following table contains allcurrently accepted Sensor Event Definitions and their respective ID. As soon as new events are added to the API, you can use their ID, even if the Definition is not yet added to this list. For more information about the API, you can have a look at the [C&S Sync API](https://cunds-syncapi.azurewebsites.net/ApiDocumentation).
+
+| ID | Definition |
+|----|------------|
+| 2 | BEDEXIT |
```
[the remark this answers is not in the document]
Comment (simon-dengler) [reply bundles/org.openhab.binding.mcd/README.md:60]: Removed the ID column

File: bundles/org.openhab.binding.mcd/README.md [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+## Channels
+
+The `mcdSensor` thing supports the following channels.  To see the sensors' events, please visit [Managing Care Digital](https://cundsdokumentation.de/) and navigate to the dashboard. 
+
+| channel     | type   | description                                   |
+|-------------|--------|-----------------------------------------------|
+| lastEvent | String | shows the last event that was sent with date and time |
+| sendEvent | String | stateless channel for sending events to the API |
+
+The channel `sendEvents` accepts valid Sensor Event Definitions as well as the corresponding ID. The following table contains allcurrently accepted Sensor Event Definitions and their respective ID. As soon as new events are added to the API, you can use their ID, even if the Definition is not yet added to this list. For more information about the API, you can have a look at the [C&S Sync API](https://cunds-syncapi.azurewebsites.net/ApiDocumentation).
```
[the remark this answers is not in the document]
Comment (simon-dengler) [reply bundles/org.openhab.binding.mcd/README.md:56]: Done

**Review by fwolter — APPROVED** [review]
LGTM
Now, you could add your binding's logo to the openHAB website. See https://www.openhab.org/docs/developer/bindings/#add-your-binding-s-logo-to-the-openhab-website

**Comment by simon-dengler** [comment]
Thanks for merging. Here is the pull request with the mcd logo:
https://github.com/openhab/openhab-docs/pull/1862

**Pull request merged**